### PR TITLE
Tesla_Act Refactor + Zap Energy Production Overhaul

### PR DESCRIPTION
**Pull request by DoctorSquishy**
# About The Pull Request

- Updates tesla arcs to a new generic zap, new defines and bitflags for such zaps. 
- energy_ball rework and updates
- SM produced zaps can also generate power now so fire up those CO2 engines
- Ground rods and tesla coils work better at grounding zaps, not 100% of the time, but better.
- Updates Grounding rods and tesla coils to TGs new energy_accumulator code (will do rad collectors separately)
- Also removes the override arg because it didn't do anything productive and nothing used it.
- Also removes a bunch of shitcode from the shock anomaly.
- Also renames 'safety' to 'SHOCK_NOGLOVES' so people actually know what it does.
- Fixes bugs with Tesla Coil delta_time handling, which caused power generation smoothing to be half as smooth as it should have been.
- Rework existing constants to be defined in terms of joules rather than in tickrate dependent units.
- Adds energy_to_joules and joules_to_energy helpers so the calculations aren't inlined everywhere.
 

## Why It's Good For The Game
Smarter tesloose code, better consistency with ground rods and tesla coils and they also have more "SOUL"

## Testing Photographs and Procedure

<details>
<summary>Tesla Power</summary>

image

</details>

<details>
<summary>Tesloose!</summary>

Less about breaching the station with explosions, but smarter at pathing to machines
image

image

</details>

<details>
<summary>Supermatter Zaps at Critical Power States</summary>

image

image

image

</details>

## PR References
- https://github.com/tgstation/tgstation/pull/48472
- https://github.com/tgstation/tgstation/pull/49260
- https://github.com/tgstation/tgstation/pull/46895
- https://github.com/tgstation/tgstation/pull/52889
- https://github.com/tgstation/tgstation/pull/53072
- https://github.com/tgstation/tgstation/pull/62144
- https://github.com/tgstation/tgstation/pull/66639
- https://github.com/tgstation/tgstation/pull/46564

## Changelog
:cl: DoctorSquishy, LemonTheDark, yoyobatty, esainane, Ghilker
add: SM produced zaps can also generate power now so fire up those CO2 engines nerds
add: Vehicles can now be zapped by both the sm and the tesla, they will shock all the people buckled to them
bug: The bicycle is no longer a perfect resistor. It will now properly get fryed when zapped by lightning.
fix: Fixes a whole bunch of issues, teslas make more sense now
tweak: Tesla is more about destroy machines than hull now, but explosions still happen, also blue light tesla
refactor: Refactored tesla_act into zap_act, it now allows for other zapping methods. Not entirely optimal, but it'll do.
refactor: Refactored supermatter_zap(), it should be a lot cleaner now
refactor: Tesla Coils, and Grounding Rods are now subtypes of a common energy_accumulator type. This centralizes logic for smoothing energy release; converting between Joules, Watts, and internal energy units; and ensures that all such accumulators display and generate power along the same smoothing curve.
qol: Tesla Coils will now display energy stored in Joules and power being produced in Watts, much like Radiation Collectors did.
qol: Grounding Rods will now display recent energy they've absorbed and the sustainable power that this energy could be producing. It should now be very obvious if you could use more Tesla Coils.
qol: supermatter zaps now change color based on the power generated, starting from red and going toward blue the higher the power emitted

/:cl:

### CODE REVIEW

File: code/__HELPERS/maths.dm [hunk 1]
```diff
@@ -94,13 +94,19 @@
...
+	return energy_units * SSmachines.wait / (1 SECONDS)
+
 /// Format an energy value measured in Power Cell units.
 /proc/display_energy(units)
 	// APCs process every (SSmachines.wait * 0.1) seconds, and turn 1 W of
 	// excess power into GLOB.CELLRATE energy units when charging cells.
 	// With the current configuration of wait=20 and CELLRATE=0.002, this
 	// means that one unit is 1 kJ.
-	return display_joules(units * SSmachines.wait * 0.1 / GLOB.CELLRATE)
+	return display_joules(energy_to_joules(units) WATTS)
```
Comment (nednaZ) [inline code/__HELPERS/maths.dm:109]: With how often this is likely to fire, making the proc here a macro or just directly putting the code in will likely help with performance.

**Review by nednaZ — CHANGES_REQUESTED** [review]
Will need test merging

File: code/__HELPERS/maths.dm (hunk 1 above)
Comment (DoctorSquishy) [reply code/__HELPERS/maths.dm:109]: The `display_energy` proc only fires on examines.

**Comment by nednaZ** [comment]
Found to not be the cause of recent issues.
Returning to test merge

**Review by nednaZ — APPROVED** [review]
Haven't been seeing any further issues after test merges.

**Pull request merged**